### ISSUE TITLE
MacOS link errors

### ISSUE THREAD

**Issue by jlblancoc** [MRPT/mrpt#923]
It FTBFS in CircleCI. 

Things to try: 
- https://stackoverflow.com/a/53040873/1631514
- https://stackoverflow.com/a/16353112/1631514

**Comment by jolting** [comment]
Weird. The compiler should treat that template as ODR-used.
```
template<class _CharT, class _Traits, class _Allocator>
basic_ostream<_CharT, _Traits>&
operator<<(basic_ostream<_CharT, _Traits>& __os,
           const basic_string<_CharT, _Traits, _Allocator>& __str)
{
    return _VSTD::__put_character_sequence(__os, __str.data(), __str.size());
}
```
I'm a bit surprised it's not inlining that function and calling `__put_character_sequence` directly.

**Comment by jlblancoc** [comment]
This is still the only failure on CircleCI.
Probably, the best way to fix it is to find the exact commit that started to fail and carefully analyze it (sigh)...

**Comment by jolting** [comment]
https://github.com/MRPT/mrpt/commit/0cbd317d8b1aeec30c525710321ef1657f22b563

**Comment by jolting** [comment]
This is probably the culprit.
https://github.com/MRPT/mrpt/commit/e2f5f387bbd7bc4e59d272f9d762e926bdfa7bef

**Comment by jolting** [comment]
I'm wondering if the `mrpt::aligned_std_basicstring<T>` in CMatrixDynamic is causing the problem.
Why isn't that just a vector?

**Comment by jlblancoc** [comment]
Thanks for searching it! So it was the big Eigen refactoring... 

> Why isn't that just a vector?

I know it should be proven by benchmarking but... it's to exploit the small string optimization. Small dynamic matrices / vectors should still fit on it so only stack is used, one less malloc/free pair.

**Comment by jlblancoc** [comment]
Tried to explicitly specify -stdlib in #930 ... no luck.

**Comment by jolting** [comment]
See if this works, https://github.com/MRPT/mrpt/tree/small_vector_optimization

**Comment by jlblancoc** [comment]
Wow, just brilliant! :-)

I guess that the OSX problem is not related at all with this type, but anyway if your patch builds, we could factor it out to its own header and use it. It's better to know for sure **what** is the small size, and not being "C++ library implementation dependant".

**Comment by jolting** [comment]
Still worth benchmarking.

**Comment by jolting** [comment]
Didn't fix the mac error.

**Comment by jolting** [comment]
Closed with #932
